gcassert: fix issues when running under bazel

healthy-pod opened the pull request:
This code change fixes issues when running gcassert under bazel by ensuring that we generate go and cgo files, and pass a path to C and C++ compilers to the lint test process.

With these changes,  `gcassert` can run as part of the `Lint` build config and doesn't need its own build config so the gcassert build config scripts are deleted.

Release note: None
Epic: CRDB-8349

Closes #65485

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103471)
<!-- Reviewable:end -->

rickystewart reviewed (commented):
What the hell is going on with the update to `dev`??

healthy-pod commented:
> What the hell is going on with the update to dev??

just testing 😄

Commit:
This code change fixes issues when running gcassert under bazel by
ensuring that we generate go and cgo files, and pass a path to C and
C++ compilers to the lint test process.

With these changes, `gcassert` can run as part of the `Lint` build
config and doesn't need its own build config so the gcassert build
config scripts are deleted.

Release note: None
Epic: CRDB-8349

Closes #65485

healthy-pod commented:
This is ready for review. Generated code and GitHub CI failures should be fixed when I merge jordanlewis/gcassert#12 and point to it instead of my fork.

rickystewart commented on build/teamcity/cockroach/ci/tests/lint_impl.sh:
```diff
@@ -2,8 +2,14 @@
 set -xeuo pipefail
 
+# GCAssert requirements -- start
+PATH="$(dirname $(bazel run @go_sdk//:bin/go --run_under=realpath)):$PATH"
```
`export PATH=`

rickystewart commented on go.mod:
```diff
@@ -159,7 +159,7 @@ require (
 	github.com/jaegertracing/jaeger v1.18.1
 	github.com/jordan-wright/email v4.0.1-0.20210109023952-943e75fe5223+incompatible
-	github.com/jordanlewis/gcassert v0.0.0-20221027203946-81f097ad35a0
+	github.com/jordanlewis/gcassert v0.0.0-20230607150025-9012d8e73bdd
```
We don't need to/shouldn't be updating `gcassert` any more, right? The `bazel` patches that were applied are all faulty and we don't seem to be using them anyway.

healthy-pod replied:
We still need to update gcassert to pull changes where we get the logs and can run in debug mode. In the current version we do not capture build logs.

rickystewart commented on pkg/cmd/generate-cgo/main.go:
```diff
@@ -148,11 +148,19 @@ import "C"
-	cppFlags := fmt.Sprintf("-I%s", filepath.Join(jemallocDir, "include"))
-	ldFlags := fmt.Sprintf("-L%s -L%s", filepath.Join(jemallocDir, "lib"), filepath.Join(projDir, "lib"))
+
+	tempToPersistentDirs := make(map[string]string, 3)
```
Nit: `srcToPersistentDirs` seems more appopriate as a name since these files don't come from "temporary" directories.

healthy-pod replied:
Done

rickystewart reviewed (changes requested):
Commit message and PR description are no longer accurate

healthy-pod commented:
> Commit message and PR description are no longer accurate
Updated

healthy-pod commented:
TFTR!

bors r=rickystewart

yuzefovich commented:
Nice work figuring this out!
